x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/RepositoryAnalyzeAction.java

```diff
@@ -730,23 +730,45 @@ public void run() {
                     return;
                 }
 
-                // complete at least request.getConcurrency() steps, but we may as well keep running for longer too
-                if (currentValue > request.getConcurrency() && otherAnalysisComplete.get()) {
-                    return;
+                if (currentValue <= request.getConcurrency() || otherAnalysisComplete.get() == false) {
+                    // complete at least request.getConcurrency() steps, but we may as well keep running for longer too
+                    transportService.sendChildRequest(
+                        nodes.get(currentValue < nodes.size() ? currentValue : random.nextInt(nodes.size())),
+                        UncontendedRegisterAnalyzeAction.NAME,
+                        new UncontendedRegisterAnalyzeAction.Request(request.getRepositoryName(), blobPath, registerName, currentValue),
+                        task,
+                        TransportRequestOptions.EMPTY,
+                        new ActionListenerResponseHandler<>(
+                            ActionListener.releaseAfter(stepListener, requestRefs.acquire()),
+                            in -> ActionResponse.Empty.INSTANCE,
+                            TransportResponseHandler.TRANSPORT_WORKER
+                        )
+                    );
+                } else {
+                    transportService.getThreadPool()
+                        .executor(ThreadPool.Names.SNAPSHOT)
+                        .execute(
+                            ActionRunnable.<Void>wrap(
+                                ActionListener.releaseAfter(
+                                    ActionListener.wrap(
+                                        r -> logger.trace("uncontended register analysis succeeded"),
+                                        AsyncAction.this::fail
+                                    ),
+                                    requestRefs.acquire()
+                                ),
+                                l -> UncontendedRegisterAnalyzeAction.verifyFinalValue(
+                                    new UncontendedRegisterAnalyzeAction.Request(
+                                        request.getRepositoryName(),
+                                        blobPath,
+                                        registerName,
+                                        currentValue
+                                    ),
+                                    repository,
+                                    l
+                                )
+                            )
+                        );
```

Comment:
Executing the action on the local node avoids having to deal with BWC. Do we need to worry about that a mixed-version cluster could give out different analysis results depending on whether the coordinating node is on the old version or on the new version?

Reply:
No, this isn't a concern for a couple of reasons: 

- we already sometimes fail the analysis for this reason because we will occasionally do the initial read of the register with a CAS(0,0):

https://github.com/elastic/elasticsearch/blob/182e7499201ea4ebfe74f39e01cf8938afeb2f29/x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/ContendedRegisterAnalyzeAction.java#L161-L169

- we also have a disclaimer in the docs that says this API might not work correctly in mixed-version clusters:

https://github.com/elastic/elasticsearch/blob/182e7499201ea4ebfe74f39e01cf8938afeb2f29/docs/reference/snapshot-restore/apis/repo-analysis-api.asciidoc#L134